pandas/tests/frame/methods/test_astype.py

```diff
@@ -726,3 +727,32 @@ def test_astype_categorical_to_string_missing(self):
         cat = df.astype("category")
         result = cat.astype(str)
         tm.assert_frame_equal(result, expected)
+
+
+class IntegerArrayNoCopy(pd.core.arrays.IntegerArray):
+    # GH 42501
+
+    @classmethod
+    def _from_sequence(cls, scalars, *, dtype=None, copy=False):
+        values, mask = coerce_to_array(scalars, dtype=dtype, copy=copy)
+        return IntegerArrayNoCopy(values, mask)
+
+    def copy(self):
+        assert False
+
+
+class Int16DtypeNoCopy(pd.Int16Dtype):
+    # GH 42501
+
+    @classmethod
+    def construct_array_type(cls):
+        return IntegerArrayNoCopy
+
+
+def test_frame_astype_no_copy():
+    # GH 42501
+    df = DataFrame({"a": [1, 4, None, 5], "b": [6, 7, 8, 9]}, dtype=object)
+    result = df.astype({"a": Int16DtypeNoCopy()}, copy=False)
+
+    tm.assert_frame_equal(df, result, check_dtype=False)
```

Comment:
this is not a great check as it doesn't assert that the original is actually unchanged (you can compare it with a propery Int16Dtype) for equality.

Reply:
@jreback I'm not totally following here.  I recognize that this call serves more as a smoke test than as a test for the specific bug fix, but I'm not clear on why we can expect the original to be unchanged since we're expressly passing copy=False to astype and the docs don't make an `inplace` style distinction for that argument. That said, the original dtype as such is unchanged:

```
>>> result.a.dtype
Int16Dtype()
>>> df.a.dtype
dtype('O')
```

Would it be sufficient to add assertions for the above attribute values?

Reply:
yes an assertion on the result dtype is ok, just want to make clear that this is not changing.